components/server/src/workspace/workspace-starter.ts

```diff
@@ -332,46 +338,123 @@ export class WorkspaceStarter {
         }
     }
 
-    private async buildImageAndStartWorkspace(
-        ctx: TraceContextWithSpan,
-        user: User,
-        workspace: Workspace,
-        instance: WorkspaceInstance,
-        envVars: ResolvedEnvVars,
-    ): Promise<StartWorkspaceResult> {
-        const { span } = ctx;
+    readonly MANAGED_PHASES: WorkspaceInstancePhase[] = ["preparing", "building", "pending"];
 
-        const forceRebuild = !!workspace.context.forceImageBuild;
-        let needsImageBuild: boolean;
-        try {
-            // if we need to build the workspace image we must not wait for actuallyStartWorkspace to return as that would block the
-            // frontend until the image is built.
-            const additionalAuth = await this.getAdditionalImageAuth(envVars);
-            needsImageBuild =
-                forceRebuild || (await this.needsImageBuild(ctx, user, workspace, instance, additionalAuth));
-            if (needsImageBuild) {
-                instance.status.conditions = {
-                    neededImageBuild: true,
-                };
+    public async start(): Promise<Disposable> {
+        this.controlAllStartingWorkspaces().catch(log.error); // Check immediately on start
+        return repeat(() => this.controlAllStartingWorkspaces(), 10 * 1000);
```

Comment:
Should we chain here, i.e. run in finally for the immediate one? Although it seems it is concurrent between different servers anyway.

Reply:
I started with simple here. As you mentioned above, we can totally turn this into a distributed job. :+1: